README.md
# Tickets API

---

## Table of Contents

- [What does it do?](#what-does-it-do)
- [Getting started](#getting-started)
- [How it works](#how-it-works)
    - [Tickets table](#tickets-table)
    - [Organisations table](#organisations-table)
- [Status and roadmap](#project-status)
- [Authors](#authors)
- [FAQ](#faq)

---
## What does it do?

The Tickets API allows you to easily create and modify tickets to distribute to your developers. Many similar tools have become overloaded with excess features. This API sticks to the basics so that you can integrate it with other popular tools such as Slack, or develop your own user interface.

---
## Getting started

1. Clone this repository.
2. Navigate to the root folder and boot up the Docker container. For more information regarding Docker, please refer to [their documentation](https://docs.docker.com/).
    ```shell
    docker-compose build
    docker-compose up
    ```
3. To run tests, navigate to the api folder.
    ```shell
    npm test
    ```
4. To see the content in the PostgreSQL database, I recommend using a GUI tool such as TablePlus. You can get it [here](https://tableplus.com/).

---
## How it works

The Tickets API uses two tables to manage its content. Both tables are accessible through their respective CRUD endpoints.

#### Tickets Table

This table is used to save all the tickets. It follows the following structure:

|uuid  |completed|summary |requirements|assigner|assignee|deadline|organisation_id|
|------|---------|--------|------------|--------|--------|--------|---------------|
|*uuid*|*boolean*|*string*|*string[]*  |*string*|*string*|*string*|*string*       |

You can modify this table by using the following endpoints:

- `GET /tickets`
    - Returns all tickets from the database.
- `GET /ticket/:uuid`
    - Returns a specific ticket by uuid.
- `POST /postTicket`
    - Saves a ticket to the database.
    - Requires a body with the following properties: 
        ```js
        {
            completed: Boolean,
            summary: String,
            requirements: String[],
            assigner: String,
            assignee: String,
            deadline: String,
            organisation_id: String
        }
        ```
- `PATCH /updateTicket/:uuid`
    - Updates **one** or **multiple** properties from a stored ticket.
    - Requires a body with the properties you wish to edit. **You do not have to provide all properties**.
        ```js
        {
            assignee: "Tim Willaert",
            deadline: "15/01/2021"
        }
        ```
- `DELETE /deleteTicket`
    - Deletes a specific ticket by uuid.
    - Requires a body which includes the uuid as a string: 
        ```js
        {
            uuid: "babf5fb0-4d18-11eb-864f-f9e4749272ab"
        }
        ```

#### Organisations Table

This table is used to save all the organisations. It follows the following structure:

|uuid  |name    |
|------|--------|
|*uuid*|*string*|

You can modify this table by using the following endpoints:

- `GET /organisations`
    - Returns all organisations from the database.
- `GET /organisations/:uuid`
    - Returns a specific organisation by uuid.
- `POST /postOrganisation`
    - Saves an organisation to the database **Work In Progress**.
    - Requires a body with the following properties: 
        ```js
        {
            name: String
        }
        ```
- `PATCH /updateOrganisation/:uuid`
    - Updates the name of a stored organisation. **Work In Progress**.
    - Requires a body with the updated name:
        ```js
        {
            name: "Erasmushogeschool Brussel"
        }
        ```
- `DELETE /deleteOrganisation`
    - Deletes a specific organisation by uuid.
    - Requires a body which includes the uuid as a string: 
        ```js
        {
            uuid: "cd8eb380-4d17-11eb-9cd9-95082ebd2d3b"
        }
        ```

---
## Project Status

This project is in development.

---
## Authors

- [Tim Willaert](https://github.com/TimWillaert) 

---
## FAQ

**Q: I'm having trouble setting up my development environment, can you help me?**

A: Please refer to [the contribution guidelines](CONTRIBUTING.md) for more information.

**Q: Can I use this project for commercial or private use?**

A: Yes! This project falls under the MIT License which allows you to do that. For more information, take a look at [the detailed license](LICENSE.md).

**Q: I have another question.**

A: Feel free to contact me at tim.willaert@student.ehb.be.


CHANGELOG.md
# Changelog

---

## 1.0.4

Features:
- API routes parameter validation

---

### 1.0.3

Features:
- checkParameters function

Fixes:
- Added function documentation
- Added API route documentation

---

### 1.0.2

Fixes:
- More detailed end to end test

---

### 1.0.1

Fixes:
- API Routes renamed

api/src/__tests__/helpers.test.js
const Helpers = require("../utils/helpers");

describe("generateUUID test", () => {
  test("check if generateUUID() generates something", () => {
    expect(Helpers.generateUUID()).not.toBeUndefined();
  });
  test("check if generated is UUID", () => {
    expect(Helpers.generateUUID()).toMatch(
      /\b[0-9a-f]{8}\b-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-\b[0-9a-f]{12}\b/
    );
  });
});

describe("checkParameters test", () => {
  test("check if function returns something", () => {
    expect(Helpers.checkParameters()).not.toBeUndefined();
  })
  test("check if function returns 400 when no parameters passed", () => {
    expect(Helpers.checkParameters()).toBe(400);
  })
  test("check if function returns 400 when parameters are wrong types", () => {
    expect(Helpers.checkParameters("hey", 0, {})).toBe(400);
  })
  test("check if function returns 400 when parameters are wrong types", () => {
    expect(Helpers.checkParameters(undefined)).toBe(400);
  })
  test("check if function returns 400 when parameters are wrong types", () => {
    expect(Helpers.checkParameters(null)).toBe(400);
  })
  test("check if function returns 400 when parameters are wrong types", () => {
    expect(Helpers.checkParameters([])).toBe(400);
  })
  test("check if function returns false when parameters are right types but wrong values", () => {
    expect(Helpers.checkParameters({hey: "hoi"}, {}, false)).toBe(false);
  })
  test("check if function returns false when parameters are right types but more parameters are not allowed", () => {
    expect(Helpers.checkParameters({name: "string"}, {name: "Tim", surname: "Willaert"}, false)).toBe(false);
  })
  test("check if function returns true when parameters are right types and more parameters are allowed", () => {
    expect(Helpers.checkParameters({name: "string"}, {name: "Tim", surname: "Willaert"}, true)).toBe(true);
  })
  test("check if function returns true when parameters are right values", () => {
    expect(Helpers.checkParameters({name: "string"}, {name: "Tim"}, false)).toBe(true);
  })
})


api/src/__tests__/endtoend.test.js
const supertest = require("supertest");
const app = require("../server");
const request = supertest(app);

describe("end to end test", () => {
    let organisationID;
    let ticketID;
    it("creates an organisation", async (done) => {
        try {
            const response = await request
                .post("/organisation")
                .send({
                    name: "Erasmushogeschool Brussel"
                });
            expect(response.status).toBe(200);
            organisationID = response.body.uuid;
            done();
        } catch (error) {}
    });
    it("updates the organisation", async (done) => {
        try {
            const response = await request
                .patch("/organisation/" + organisationID)
                .send({name: "EhB"});
            expect(response.status).toBe(200);
            done();
        } catch (error) {}
    })
    it("creates a ticket", async (done) => {
        try {
            const response = await request
                .post("/ticket")
                .send({ 
                summary: "We need to update the visuals of our website",
                requirements: [
                    "Update landing page banner image",
                    "Update color scheme to new branding",
                    "Update footer copyright to 2021"
                ],
                assigner: "Tim Willaert",
                assignee: undefined,
                deadline: "11/01/2021",
                organisation_id: organisationID
            });
            expect(response.status).toBe(200);
            ticketID = response.body.uuid;
            done();
        } catch (error) {}
    })
    it("assigns a developer to ticket", async (done) => {
        try {
            const response = await request
                .patch("/ticket/" + ticketID)
                .send({assignee: "Chiel Habils"});
            expect(response.status).toBe(200);
            done();
        } catch (error) {}
    })
    it("marks the ticket as complete", async (done) => {
        try {
            const response = await request
                .patch("/ticket/" + ticketID)
                .send({completed: true});
            expect(response.status).toBe(200);
            done();
        } catch (error) {}
    })
    it("deletes the ticket", async (done) => {
        try {
            const response = await request
                .delete("/ticket")
                .send({uuid: ticketID});
            expect(response.status).toBe(200);
            done();
        } catch (error) {}
    })
    it("deletes the organisation", async (done) => {
        try {
            const response = await request
                .delete("/organisation")
                .send({uuid: organisationID});
            expect(response.status).toBe(200);
            done();
        } catch (error) {}
    })
});

api/src/__tests__/integration.test.js
const supertest = require("supertest");
const app = require("../server");
const request = supertest(app);

describe("GET /tickets", () => {
  it("responds with 200 and returns all tickets", async (done) => {
    try {
      const response = await request.get("/tickets");
      expect(response.status).toBe(200);
      expect(typeof response.body).toBe("object", done());
    } catch (error) {}
  });
});

describe("Create, find and update ticket", () => {
  let uuid;
  it("responds with 200 if ticket is added to db", async (done) => {
    try {
      const response = await request
        .post("/ticket")
        .send({ 
          summary: "We need to update the visuals of our website",
          requirements: [
            "Update landing page banner image",
            "Update color scheme to new branding",
            "Update footer copyright to 2021"
          ],
          assigner: "Tim Willaert",
          deadline: "11/01/2021",
          organisation_id: 1
       });
      expect(response.status).toBe(200);
      uuid = response.body.uuid;
      done();
    } catch (error) {}
  });
  it("finds the recently added record in db", async (done) => {
    try {
      const response = await request.get("/ticket/" + uuid);
      expect(response.status).toBe(200);
      expect(typeof response.body).toBe("object");
      done();
    } catch (error) {}
  });
  it("responds with 200 if ticket is updated", async (done) => {
    try {
      const response = await request
        .patch("/ticket/" + uuid)
        .send({deadline: "13/01/2021"});
      expect(response.status).toBe(200);
      done();
    } catch (error) {}
  })
});

describe("DELETE /deleteTicket", () => {
  it("responds with 400 if no uuid is provided", async (done) => {
    try {
      const response = await request.delete("/ticket").send({});
      expect(response.status).toBe(400);
      done();
    } catch (error) {}
  });
});


api/src/utils/helpers.js
const { v1: uuidv1 } = require("uuid");
const sortObject = require('sort-object-keys');

const Helpers = {
  /**
  * generateUUID generates a v1 UUID
  * @param: none
  * @returns: v1 UUID
  */
  generateUUID: () => {
    const uuid = uuidv1();
    return uuid;
  },

  /**
  * checkParameters checks if all expected parameters are present and of the right type
  * @param:
  * - expectedParameters: object (list of expected properties and their type)
  *   - for example: {uuid: 'string', name: 'string'}
  * - givenParameters: object (list of given parameters and their value)
  *   - for example: {uuid: 'e89a4df0-58df-11eb-9bcd-e7a81d580945', name: 'Tim}
  * - allowOtherParameters: boolean (allow more parameters to be present than the expected ones)
  * @returns: 
  * - 400 if not all 3 parameters are present or their types are wrong
  * - true if all expected parameters are present and of the right type
  * - false if not all expected parameters are present and/or of the wrong type
  */
  checkParameters: (expectedParameters, givenParameters, allowOtherParameters) => {
    if(
      typeof expectedParameters == "object" &&
      typeof givenParameters == "object" &&
      typeof allowOtherParameters == "boolean"
    ) {
      let sortedExpectedParams = sortObject(expectedParameters);
      let sortedGivenParams = sortObject(givenParameters);
      let matchingParameters = 0;
      for(const requiredParam in sortedExpectedParams){
        if(
          requiredParam in sortedGivenParams &&
          typeof sortedGivenParams[requiredParam] == expectedParameters[requiredParam]
        ) {
          matchingParameters++;
        }
      }
      if(Object.keys(sortedGivenParams).length > matchingParameters && allowOtherParameters == false){
        return false;
      } else if(Object.keys(sortedGivenParams).length > matchingParameters && allowOtherParameters == true){
        return true;
      } else if(matchingParameters < Object.keys(sortedExpectedParams).length){
        return false;
      } else if(matchingParameters == Object.keys(sortedExpectedParams).length){
        return true;
      } else{
        return false;
      }
    }
    return 400;
  },
};

module.exports = Helpers;


api/src/utils/DatabaseHelper.js
const Helpers = require("../utils/helpers.js");

const pg = require("knex")({
    client: "pg",
    version: "9.6",
    searchPath: ["knex", "public"],
    connection: process.env.PG_CONNECTION_STRING
      ? process.env.PG_CONNECTION_STRING
      : "postgres://example:example@localhost:5432/test",
  });

const DatabaseHelper = {
    async initialiseTables() {
        await pg.schema.hasTable("tickets").then(async (exists) => {
          if (!exists) {
            await pg.schema
              .createTable("tickets", (table) => {
                table.increments();
                table.uuid("uuid");
                table.boolean("completed").defaultTo(false);
                table.string("summary");
                table.specificType("requirements", "text ARRAY");
                table.string("assigner");
                table.string("assignee");
                table.string("deadline");
                table.string("organisation_id");
                table.timestamps(true, true);
              })
          }
        });
        await pg.schema.hasTable("organisations").then(async (exists) => {
          if (!exists) {
            await pg.schema
              .createTable("organisations", (table) => {
                table.increments();
                table.uuid("uuid");
                table.string("name");
                table.timestamps(true, true);
              })
              .then(async () => {
                console.log("created table organisations");
                for (let i = 0; i < 10; i++) {
                  const uuid = Helpers.generateUUID();
                  await pg
                    .table("organisations")
                    .insert({ uuid, name: `Organisation ${i}`});
                }
              });
          }
        });
    }
}

module.exports = DatabaseHelper, pg;

api/src/server.js
const express = require("express");
const bodyParser = require("body-parser");
const http = require("http");
const Helpers = require("./utils/helpers.js");
const DatabaseHelper = require("./utils/DatabaseHelper.js");
const { send } = require("process");

const pg = require("knex")({
  client: "pg",
  version: "9.6",
  searchPath: ["knex", "public"],
  connection: process.env.PG_CONNECTION_STRING
    ? process.env.PG_CONNECTION_STRING
    : "postgres://example:example@localhost:5432/test",
});

const app = express();
http.Server(app);

app.use(bodyParser.json());
app.use(
  bodyParser.urlencoded({
    // to support URL-encoded bodies
    extended: true,
  })
);

/******************* */
/* TICKETS ENDPOINTS */
/******************* */

/**
* GET /tickets
* @param: none
* @returns: all ticket records from the database
*/
app.get("/tickets", async (req, res) => {
  const result = await pg
    .select(["uuid", "summary", "requirements", "completed", "assigner", "assignee", "deadline", "organisation_id", "created_at"])
    .from("tickets");
  res.status(200).send(result);
});

/**
* GET /ticket/uuid
* @param: uuid
* @returns: a specific ticket record from the database by uuid
*/
app.get("/ticket/:uuid", async (req, res) => {
  if(Helpers.checkParameters({uuid: 'string'}, {uuid: req.params.uuid}, false)){
    const result = await pg
      .select(["uuid", "summary", "requirements", "completed", "assigner", "assignee", "deadline", "organisation_id", "created_at"])
      .from("tickets")
      .where({ uuid: req.params.uuid });
    if(result.length == 0){
      res.status(404).send("Ticket not found");
    } else{
      res.status(200).send(result[0]);
    }
  } else{
    res.status(400).send("Invalid parameters");
  }
});

/**
* POST /ticket
* @param: {uuid, summary, requirements, assigner, assignee, deadline, organisation_id}
* @returns: the uuid of the added ticket
*/
app.post("/ticket", (req, res) => {
  let uuid = Helpers.generateUUID();
  let expectedParams = {
    uuid: 'string',
    summary: 'string',
    requirements: 'array',
    assigner: 'string',
    assignee: 'string',
    deadline: 'string',
    organisation_id: 'string'
  }
  if(Helpers.checkParameters(expectedParams, {
    uuid: uuid,
    summary: req.body.summary,
    requirements: req.body.requirements,
    assigner: req.body.assigner,
    assignee: req.body.assignee,
    deadline: req.body.deadline,
    organisation_id: req.body.organisation_id,
    created_at: new Date()
  }, true)){
    pg.insert({
      uuid: uuid,
      summary: req.body.summary,
      requirements: req.body.requirements,
      assigner: req.body.assigner,
      assignee: req.body.assignee,
      deadline: req.body.deadline,
      organisation_id: req.body.organisation_id,
      created_at: new Date(),
    })
      .into("tickets")
      .then(() => {
        res.json({ uuid: uuid });
    });
  } else{
    res.status(400).send("Invalid parameters");
  }
});

/**
* PATCH /ticket/:uuid
* @param: object with the properties that need to be updated/patched
* @returns: status code 200
*/
app.patch("/ticket/:uuid", async (req, res) => {
  if(Helpers.checkParameters({uuid: 'string'}, req.body, true)){
    pg('tickets')
      .where({uuid: req.params.uuid})
      .update(req.body)
      .then(() => {
        res.sendStatus(200);
      })
  } else {
    res.status(400).send("Invalid parameters");
  }
});

/**
* DELETE /ticket
* @param: {uuid}
* @returns: status code 200
*/
app.delete("/ticket", (req, res) => {
  if(Helpers.checkParameters({uuid: 'string'}, {uuid: req.body.uuid}, false)){
    pg('tickets')
      .where({ uuid: req.body.uuid })
      .del()
      .then(() => {
        res.sendStatus(200);
    })
  } else {
    res.status(400).send("Invalid parameters");
  }
});

/************************* */
/* ORGANISATIONS ENDPOINTS */
/************************* */

/**
* GET /organisations
* @param: none
* @returns: all organisation records from the database
*/
app.get("/organisations", async (req, res) => {
  const result = await pg
    .select(["uuid", "name", "created_at"])
    .from("organisations");
  res.status(200).send(result);
});

/**
* GET /organisation/uuid
* @param: uuid
* @returns: a specific organisation record from the database by uuid
*/
app.get("/organisation/:uuid", async (req, res) => {
  if(Helpers.checkParameters({uuid: 'string'}, {uuid: req.params.uuid}, false)){
    const result = await pg
      .select(["uuid", "name", "created_at"])
      .from("organisations")
      .where({ uuid: req.params.uuid });
    if(result.length == 0){
      res.status(404).send("Organisation not found");
    } else{
      res.status(200).send(result[0]);
    }
  } else {
    res.status(400).send("Invalid parameters");
  }
});

/**
* POST /organisation
* @param: {uuid, name}
* @returns: the uuid of the added organisation
*/
app.post("/organisation", (req, res) => {
  let uuid = Helpers.generateUUID();
  if(Helpers.checkParameters({uuid: 'string', name: 'string'}, {uuid: uuid, name: req.body.name, created_at: new Date()}, true)){
    pg.insert({
      uuid: uuid,
      name: req.body.name,
      created_at: new Date(),
    })
      .into("organisations")
      .then(() => {
        res.json({ uuid: uuid });
    });
  } else {
    res.status(400).send("Invalid parameters");
  }
});

/**
* PATCH /organisation/:uuid
* @param: object with the properties that need to be updated/patched
* @returns: status code 200
*/
app.patch("/organisation/:uuid", async (req, res) => {
  if(Helpers.checkParameters({uuid: 'string'}, {uuid: req.params.uuid}, true)){
    pg('organisations')
      .where({uuid: req.params.uuid})
      .update(req.body)
      .then(() => {
        res.sendStatus(200);
    })
  } else {
    req.status(400).send("Invalid parameters");
  }
});

/**
* DELETE /organisation
* @param: {uuid}
* @returns: status code 200
*/
app.delete("/organisation", (req, res) => {
  if(Helpers.checkParameters({uuid: 'string'}, {uuid: req.body.uuid}, false)){
    pg('organisations')
      .where({ uuid: req.body.uuid })
      .del()
      .then(() => {
        res.sendStatus(200);
    })
  } else{
    res.status(400).send("Invalid parameters");
  }
});

DatabaseHelper.initialiseTables();

module.exports = app;
